helix-core/src/syntax.rs

```diff
@@ -1946,6 +2098,116 @@ impl<'a> Iterator for HighlightIter<'a> {
     }
 }
 
+impl<'a> Iterator for RainbowIter<'a> {
+    type Item = Result<HighlightEvent, Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        'main: loop {
+            // If we've already determined the next highlight boundary, just return it.
+            if let Some(e) = self.next_event.take() {
+                return Some(Ok(e));
+            }
+
+            // If the iterator has been cancelled, return an Error.
+            if let Err(err) = self.check_cancellation() {
+                return Some(Err(err));
+            }
+
+            // If none of the layers have any more highlight boundaries, terminate.
+            if self.layers.is_empty() {
+                return None;
+            }
+
+            if let Some(event) = self.scan_to_earliest_capture() {
+                return event;
+            };
+
+            let layer = &mut self.layers[0];
+            let (match_, capture_index) = layer.captures.next().unwrap();
+            let capture = match_.captures[capture_index];
+            let range = capture.node.byte_range();
```

Comment:
note: this is fully duplicate but doesn't lend it self well to moving to a seperate function as it's mostly controlflow that's why I suggested merging both iterator implementations